File: macosx/WallpaperInfoApp/WallpaperInfoApp/ImageFileManager.swift
//
//  ImageFileManager.swift
//  WallpaperInfoApp
//
//  Created by Brad Park on 4/12/20.
//  Copyright © 2020 Brad Park. All rights reserved.
//

import Foundation

class ImageFileManager {
	
    init() {
        resetFiles()
    }

    func resetFiles() {
        objc_sync_enter(self._fileLock)
		_themeReady = false
		_unclassifiedPaths = WLinkedHashMap<String, String>() // will be overridden
        _wallpaperPaths = WLinkedHashMap<String, String>() // will be overridden
		_themePaths = WLinkedHashMap<String, String>()
		_unthemePaths = WLinkedHashMap<String, String>()
		_usedPaths = WLinkedHashMap<String, String>()
		_lastUsedPaths = WLinkedHashMap<String, String>()
		let wpsi = WallpaperServiceInfo.getInstance()
		wpsi.setLastUsedPaths(_lastUsedPaths);
        objc_sync_exit(self._fileLock)
        stopWatching()
    }
    
    func stopWatching() {
        _writeInterrupt = true
        waitForListFileWriting()
        objc_sync_enter(_fileLock)
        if (_fileObserver != nil) {
            _fileObserver.stopWatching()
            _fileObserver = nil
        }
        objc_sync_exit(_fileLock)
    }
    
    func setSourceRootPath(_ sourceRootPath: String) {
        resetFiles()
        _writeInterrupt = false
        _sourceRootPath = sourceRootPath
		if (!BPUtil.fileExists(sourceRootPath)) {
			return
		}
        if (readListCache()) {
            objc_sync_enter(_fileLock)
            BPUtil.BPLog("ImageFileManager.readListCache.");
            WPath.setPlatformRootPath(path: _unclassifiedPaths.getWPath(at: 0).path)
            classifySourceByTheme()
            objc_sync_exit(_fileLock)
            _themeReady = true
        }
        _fileObserver = PlatformFileObserver(path:_sourceRootPath)
        _fileObserver.onEvent =  { event, wppath in
            switch event {
                case PlatformFileObserver.ADD:
					self.addWPath(wppath)
                    break
                case PlatformFileObserver.REMOVE:
					self.removePath(path:wppath.path)
                    break
                default:
                    return
            }
        }
        _fileObserver.fileWatchingStarted = { (sourcePaths) in
            BPUtil.BPLog("ImageFileManager.fileWatchingStarted.")
			self._wallpaperPaths = sourcePaths
            self.writeListCache(paths:self._wallpaperPaths, listCacheFile: self._wallpaperListCacheFile)
            objc_sync_enter(self._fileLock)
            // compare count to determine to use it over cache list.
            if (self._wallpaperPaths.count > self.getTotalImageCount()) {
                self.replaceCacheWithRealPaths(realPaths: self._wallpaperPaths)
            }
			objc_sync_exit(self._fileLock)
        }
		_fileObserver.exifReadFinished = { (exifFilePaths) in
            BPUtil.BPLog("ImageFileManager.exifReadFinished.")
            self._unclassifiedPaths = exifFilePaths
            self.writeListCache(paths:self._unclassifiedPaths, listCacheFile: self._photoListCacheFile)
            objc_sync_enter(self._fileLock)
            if (self._usingListCache) {
                self.replaceCacheWithRealPaths(realPaths: self._unclassifiedPaths, realPaths2: self._wallpaperPaths)
            } else {
                self.classifySourceByTheme()
            }
			objc_sync_exit(self._fileLock)
        }
        // it's background scanning. _themeReady flag will protect to access the file source during scanning
        // It will call fileWatchingStarted when it's done
        BPUtil.BPLog("ImageFileManager.setSourceRootPath at " + sourceRootPath)
        _fileObserver.startWatching()
    }
    
    private func getRandomPathFromSource() -> WPath? {
		if (_themePaths == nil || _usedPaths == nil) {
            BPUtil.BPLog("_sourcePaths shouldn't be nil.")
			return nil
		}
		objc_sync_enter(_fileLock)
		if (_themePaths.count <= 0) {
			// Rewind from the beginning
			rewindSource()
			if (_themePaths.count <= 0) {
				// when there is no images;
                BPUtil.BPLog("Error: There is no theme image in entire source.");
				objc_sync_exit(_fileLock)
				return nil
			}
		}
		// pick new image path from source
		let index = Int.random(in: 0..<_themePaths.count);
		let newWPath = _themePaths.getWPath(at: index)
		// remove from source path
		_themePaths.removeValueForKey(key: newWPath.path)
		// (in case source is small) Check if new path is still in last used path, then remove it to avoid the loop.
		_lastUsedPaths.removeValueForKey(key: newWPath.path)
		// add to last used path
		_lastUsedPaths.put(value: newWPath.exif, forKey: newWPath.path);
		// mainatain size for _lastUsedPaths
		if (_lastUsedPaths.count > WallpaperService._maxLastUsedPaths) {
			_lastUsedPaths.remove(at: 0)
		}
		// add to used path
		_usedPaths.put(value: newWPath.exif, forKey: newWPath.path)
		objc_sync_exit(_fileLock)
		return newWPath
    }
    
    func retrievePathFromSource(pivotWPath:WPath?, offset:Int) -> WPath? {
		if (pivotWPath == nil || pivotWPath!.path == "" ) {
			return getRandomPathFromSource()
		}
		if (_lastUsedPaths == nil) {
            BPUtil.BPLog("_lastUsedPaths shouldn't be nil.")
			return nil
		}
		objc_sync_enter(_fileLock)
		let index = _lastUsedPaths.firstIndex(of: pivotWPath!.path)
		if (index == nil) {                                        // No matching image
			if (offset == -1) {
                BPUtil.BPLog("shouldn't happen when no index but offset is -1")
			}
			objc_sync_exit(_fileLock)
			return getRandomPathFromSource();
		} else if (index! + offset < 0) {                        // get previous at the first
			objc_sync_exit(_fileLock)
			return _lastUsedPaths.getWPath(at: 0)
		} else if (index! + offset < _lastUsedPaths.count) {     // get within the range
			objc_sync_exit(_fileLock)
			return _lastUsedPaths.getWPath(at: index! + offset)
		}
		objc_sync_exit(_fileLock)
		return getRandomPathFromSource();                     // get next at the last
    }
    
	private func addWPath(_ wppath: WPath) {
		if (_themePaths == nil) {
            BPUtil.BPLog("_sourcePaths shouldn't be None.")
			return;
		}
		objc_sync_enter(_fileLock)
		let wpsi = WallpaperServiceInfo.getInstance()
		let themeInfo = wpsi.getThemeInfo()
		if (themeInfo.isThemeImage(coden: wppath.coden())) {
			_themePaths.put(value: wppath.exif, forKey: wppath.path);
		} else {
			_unthemePaths.put(value: wppath.exif, forKey: wppath.path);
		}
		objc_sync_exit(_fileLock)
    }
    
    private func removePath(path: String) {
        if (_themePaths == nil || _usedPaths == nil) {
            BPUtil.BPLog("_sourcePaths shouldn't be None.")
            return;
        }
		objc_sync_enter(_fileLock)
		_lastUsedPaths.removeValueForKey(key: path)
		_themePaths.removeValueForKey(key: path)
		_unthemePaths.removeValueForKey(key: path)
		_usedPaths.removeValueForKey(key: path)
		objc_sync_exit(_fileLock)
    }
    
    func classifySourceByTheme() {
        let wpsi = WallpaperServiceInfo.getInstance()
        wpsi.getThemeInfo().classifyPathsByTheme(
            source: _unclassifiedPaths,
            theme: &_themePaths,
            untheme: &_unthemePaths)
    }
	
	private func rewindSource() {
		_themeReady = false
        BPUtil.BPLog("rewindSource is called.")
        _unclassifiedPaths = _usedPaths
        _usedPaths = WLinkedHashMap<String, String>()
        classifySourceByTheme()
		_themeReady = true
	}
	
	func prepareSourceForTheme() {
		_themeReady = false
		let wpsi = WallpaperServiceInfo.getInstance()
        let requestedThemeInfo = wpsi.getThemeInfo().copy() as! ThemeInfo
        BPUtil.BPLog("prepareSourceForTheme is called with " + requestedThemeInfo._theme.stringValue)
		DispatchQueue.background(delay: 1.1, completion:{
            //print("prepareSourceForTheme enters dispatchqueue with " + requestedThemeInfo._theme.stringValue)
			objc_sync_enter(self._fileLock)
			//print("prepareSourceForTheme enters _fileLock with " + requestedThemeInfo._theme.stringValue)
			if (requestedThemeInfo.equals(wpsi.getThemeInfo())) {
				//print("prepareSourceForTheme execution starts with " + requestedThemeInfo._theme.stringValue)
				self._unclassifiedPaths = self._themePaths
				let backupPaths = self._unthemePaths
				self._themePaths = WLinkedHashMap<String, String>()
				self._unthemePaths = WLinkedHashMap<String, String>()
				self.classifySourceByTheme()
				self._unclassifiedPaths = backupPaths
				self.classifySourceByTheme()
				self._themeReady = true
                BPUtil.BPLog("prepareSourceForTheme is done with " + requestedThemeInfo._theme.stringValue)
			}
			objc_sync_exit(self._fileLock)
		})
	}
    
    func isThemeReady() -> Bool {
        return _themeReady
    }
	
    func getImageStat(path: String) -> String {
		var index = _usedPaths.count + _unthemePaths.count
		objc_sync_enter(_fileLock)
		if let i = _lastUsedPaths.firstIndex(of: path) {  // current image is in the mid offset
			index = index - _lastUsedPaths.count + i
		}
		objc_sync_exit(_fileLock)
		return String(index + 1)  + " of " + String(getTotalImageCount())
    }
    
    func getTotalImageCount() -> Int {
		objc_sync_enter(_fileLock)
		let count = _usedPaths.count + _themePaths.count + _unthemePaths.count
		objc_sync_exit(_fileLock)
		return count
    }
	
    static func getDefaultSourceRootPath() -> String {
		return NSHomeDirectory() + "/Pictures"
    }

    private func readListCache() -> Bool {
        _usingListCache = false;
        _unclassifiedPaths.removeAll(keepCapacity: 0)
       if (BPUtil.fileExists(_wallpaperListCacheFile)) {
           BPUtil.BPLog("readListCache: reading wallpaper list cache file at \(_wallpaperListCacheFile)")
           waitForListFileWriting()
           if let reader = StreamReader(url: URL(fileURLWithPath:_wallpaperListCacheFile)) {
               if let path = reader.nextLine() {
                   if (path == _sourceRootPath) {
                       var line = reader.nextLine()
                       while(line != nil) {
                           let wpathInFile = line!.components(separatedBy: "\t")
                           _unclassifiedPaths.put(value: "", forKey: wpathInFile[0])
                           line = reader.nextLine()
                       }
                   }
               }
           }
       } else {
           BPUtil.BPLog("readListCache: no wallpaper list cache file")
       }
        if (BPUtil.fileExists(_photoListCacheFile)) {
            BPUtil.BPLog("readListCache: reading photo list cache file at \(_photoListCacheFile)")
            waitForListFileWriting()
            if let reader = StreamReader(url: URL(fileURLWithPath:_photoListCacheFile)) {
                if let path = reader.nextLine() {
                    if (path == _sourceRootPath) {
                        var line = reader.nextLine()
                        while(line != nil) {
                            let wpathInFile = line!.components(separatedBy: "\t")
                            let exif = wpathInFile.count > 1 ? wpathInFile[1] : ""
                            _unclassifiedPaths.put(value: exif, forKey: wpathInFile[0])
                            line = reader.nextLine()
                        }
                    }
                }
            }
        } else {
            BPUtil.BPLog("readListCache: no photo list cache file")
        }
       _usingListCache = _unclassifiedPaths.count > 0;
        return _usingListCache;
    }

    // nil to delete
    private func writeListCache(paths:WLinkedHashMap<String, String>?, listCacheFile:String) {
        if (paths != nil) {
            waitForListFileWriting()
            _writeFileGroup = DispatchGroup()
            _writeFileGroup!.enter()
            DispatchQueue.global(qos: .background).async {
                do {
                    BPUtil.BPLog("writeListCache \(listCacheFile) started")
                    let listCacheFileUrlTemp = URL(fileURLWithPath:listCacheFile+".tmp")
                    guard let outputStream = OutputStream(url: listCacheFileUrlTemp, append: false) else { return }
                    outputStream.open()
                    try outputStream.write(self._sourceRootPath + "\n")
                    var iter = paths!.makeIterator()
                    while let wpath = iter.nextWPath() {
                        if (self._writeInterrupt) {
                            BPUtil.BPLog("writeListCache is interrupted for \(listCacheFile)")
                            break
                        }
                        try outputStream.write(wpath.path + "\t" + wpath.exif + "\n")
                    }
                    outputStream.close()
                    if (!self._writeInterrupt) {
                        if (BPUtil.fileExists(listCacheFile)) {
                            try FileManager.default.removeItem(atPath: listCacheFile)
                        }
                        try FileManager.default.moveItem(atPath: listCacheFile + ".tmp", toPath: listCacheFile)
                    } else {
                        try FileManager.default.removeItem(at: listCacheFileUrlTemp)
                    }
                    BPUtil.BPLog("writeListCache \(listCacheFile) ended")
                } catch {
                    print(error)
                    return
                }
                self._writeFileGroup!.leave()
            }
        }
    }
    
    private func waitForListFileWriting() {
        if let writeFileGroup = _writeFileGroup {
            BPUtil.BPLog("waitForListFileWriting started")
            writeFileGroup.wait()
            BPUtil.BPLog("waitForListFileWriting ended")
        }
    }

    private func replaceCacheWithRealPaths(realPaths: WLinkedHashMap<String, String>, realPaths2: WLinkedHashMap<String, String>? = nil) {
         _themeReady = false;
         _unclassifiedPaths = realPaths;
        _themePaths.removeAll(keepCapacity: 0)
         _unthemePaths.removeAll(keepCapacity: 0)
         classifySourceByTheme()
        if (realPaths2 != nil) {
            _unclassifiedPaths = realPaths2;
            classifySourceByTheme();
        }
        var iter = _usedPaths.makeIterator()
        while let wpath = iter.nextWPath() {
            // remove from _themePaths which was already used from cache
            _themePaths.removeValueForKey(key: wpath.path)
        }
         _usingListCache = false
         _themeReady = true
         WallpaperWidgetProvider.updateLabelWidget(text: "Image list is updated.")
    }
    
    private var _fileObserver: PlatformFileObserver!
    private var _unclassifiedPaths: WLinkedHashMap<String, String>!
    private var _wallpaperPaths: WLinkedHashMap<String, String>!
    private var _themePaths: WLinkedHashMap<String, String>!
    private var _unthemePaths: WLinkedHashMap<String, String>!
    private var _usedPaths: WLinkedHashMap<String, String>!
    private var _lastUsedPaths: WLinkedHashMap<String, String>!
    private var _sourceRootPath: String!
	private var _themeReady = false
    private var _usingListCache = false
    private var _writeInterrupt = false
    private var _writeFileGroup:DispatchGroup? = nil
    private let _wallpaperListCacheFile = BPUtil.getHomeDirectory() + "/Documents/.wallpaper_list.txt"
    private let _photoListCacheFile = BPUtil.getHomeDirectory() + "/Documents/.photo_list.txt"
    // only for Mac OS X
    private var _fileLock = NSObject()
}

typealias WLinkedHashMap = LinkedHashMap
extension WLinkedHashMap {
	func getWPath(at index:Int) -> WPath {
		if let key = self.keys[index] as? String {
			if let value = self[self.keys[index]] as? String {
				return WPath(path: key, exif: value)
			}
		}
		assertionFailure("Error - Fail to get WPath from WLinkedHashMap")
		return WPath(path: "", exif: "")
	}
}

extension LinkedHashMapIterator {
    mutating func nextWPath() -> WPath? {
        if let (k, v) = next() {
            return WPath(path:k as! String, exif:v as! String)
        } else {
            return nil
        }
    }
}

File: apps/desktop-ubuntu/WallpaperInfoApp/ImageFileManager.py
#!/usr/bin/env python3

from collections import OrderedDict
import random
#import os.path
import os
from threading import Thread
from threading import Timer

import WallpaperServiceInfo as w
import WallpaperService
import WPath
import BPUtil
import PlatformFileObserver
import PlatformInfo
import WallpaperWidgetProvider

class ImageFileManager:

    def __init__(self):
        self._lock = PlatformInfo.BPLock()
        self._fileObserver = None
        self._usingListCache = False
        self._writeInterrupt = False
        self._writeFileThread = None
        self.resetFiles()

    def resetFiles(self):
        self._lock.acquire()
        self._themeReady = False
        self._unclassifiedPaths = OrderedDict() # will be overriden
        self._wallpaperPaths = OrderedDict() # will be overriden
        self._themePaths = OrderedDict()
        self._unthemePaths = OrderedDict()
        self._usedPaths = OrderedDict()
        self._lastUsedPaths = OrderedDict()
        wpsi = w.WallpaperServiceInfo()
        wpsi.setLastUsedPaths(self._lastUsedPaths)
        self._lock.release()
        self.stopWatching()

    def stopWatching(self):
        self._writeInterrupt = True
        self.waitForListFileWriting()
        self._lock.acquire()
        if (self._fileObserver != None):
            self._fileObserver.stopWatching()
            self._fileObserver = None
        self._lock.release()

    def setSourceRootPath(self, sourceRootPath):
        self.resetFiles()
        self._writeInterrupt = False
        self._sourceRootPath = sourceRootPath
        if not BPUtil.fileExists(sourceRootPath):
            BPUtil.BPLog(sourceRootPath + "directory doesn't exist.")
            return
        if (self.readListCache()):
            self._lock.acquire()
            BPUtil.BPLog("ImageFileManager.readListCache.")
            WPath.WPath.setPlatformRootPath(list(self._unclassifiedPaths.keys())[0])
            self.classifySourceByTheme()
            self._lock.release()
            self._themeReady = True
        self._fileObserver = PlatformFileObserver.PlatformFileObserver(self._sourceRootPath, self)
        BPUtil.BPLog("ImageFileManager.setSourceRootPath at " + sourceRootPath)
        self._fileObserver.startWatching()

    def onEvent(self, event, wpath):
        if event == PlatformFileObserver.ADD:
            self.addWPath(wpath)
        elif event == PlatformFileObserver.REMOVE:
            self.removePath(wpath.path)
        else:
            return

    def fileWatchingStarted(self, sourcePaths):
        BPUtil.BPLog("ImageFileManager.fileWatchingStarted.")
        self._wallpaperPaths = sourcePaths
        self.writeListCache(self._wallpaperPaths, _wallpaperListCacheFile)
        self._lock.acquire()
        # compare count to determine to use it over cache list.
        if (len(self._wallpaperPaths) > self.getTotalImageCount()):
            self.replaceCacheWithRealPaths(self._wallpaperPaths)
        self._lock.release()
        
    def exifReadFinished(self, exifPaths):
        BPUtil.BPLog("ImageFileManager.exifReadFinished.")
        self._unclassifiedPaths = exifPaths
        self.writeListCache(self._unclassifiedPaths, _photoListCacheFile)
        self._lock.acquire()
        if (self._usingListCache):
            self.replaceCacheWithRealPaths(self._unclassifiedPaths, self._wallpaperPaths)
        else:
            self.classifySourceByTheme()
        self._lock.release()
		
    def getRandomPathFromSource(self):
        if self._themePaths == None or self._usedPaths == None:
            BPUtil.BPLog("_sourcePaths shouldn't be None.")
            return None
        self._lock.acquire()
        if len(self._themePaths) <= 0:
            # Rewind from the beginning
            self.rewindSource()
            if len(self._themePaths) <= 0:
                # when there is no images
                BPUtil.BPLog("There is no theme image in entire source.")
                self._lock.release()
                return None
        # pick new image path from source
        path = random.choice(list(self._themePaths.keys()))
        exif = self._themePaths[path]
        # remove from source path
        if path in self._themePaths:
            del self._themePaths[path]
        # (in case source is small) Check if new path is still in last used path, then remove it to avoid the loop.
        if path in self._lastUsedPaths:
            del self._lastUsedPaths[path]
        # add to last used path
        self._lastUsedPaths[path]= exif
        # mainatain size for self._lastUsedPaths
        if len(self._lastUsedPaths) > WallpaperService._maxLastUsedPaths:
            self._lastUsedPaths.popitem(last=False)
        # add to used path
        self._usedPaths[path] = exif
        self._lock.release()
        return WPath.WPath(path, exif)
            
    def retrievePathFromSource(self, pivotWPath, offset):
        if pivotWPath == None:
            return self.getRandomPathFromSource()
        if (self._lastUsedPaths == None):
            BPUtil.BPLog("self._lastUsedPaths shouldn't be None.")
            return None
        self._lock.acquire()
        index = list(self._lastUsedPaths.keys()).index(pivotWPath.path)
        if index < 0:                                      # No matching image
            if offset == -1:
                BPUtil.BPLog("shouldn't happen when no index but offset is -1")
                wpath = self.getRandomPathFromSource()
        elif index + offset < 0:                           # get previous at the first
            path = list(self._lastUsedPaths.keys())[0]
            wpath = WPath.WPath(path, self._lastUsedPaths[path])
        elif index + offset < len(self._lastUsedPaths):    # get within the range
            path = list(self._lastUsedPaths.keys())[index + offset]
            wpath = WPath.WPath(path, self._lastUsedPaths[path])
        else:
            self._lock.release()
            return self.getRandomPathFromSource() # get next at the last
        self._lock.release()
        return wpath                 

    def addWPath(self, wpath):
        if (self._themePaths == None):
            BPUtil.BPLog("_sourcePaths shouldn't be None.")
            return
        self._lock.acquire()
        wpsi = w.WallpaperServiceInfo()
        themeInfo = wpsi.getThemeInfo()
        if themeInfo.isThemeImage(wpath.coden()):
            self._themePaths[wpath.path] = wpath.exif
        else:
            self._unthemePaths[wpath.path] = wpath.exif
        self._lock.release()
				            		
    def removePath(self, path):
        if self._themePaths == None or self._usedPaths == None:
            BPUtil.BPLog("_sourcePaths shouldn't be None.")
            return
        self._lock.acquire()
        if path in self._lastUsedPaths:
            del self._lastUsedPaths[path]
        if path in self._themePaths:
            del self._themePaths[path]
        if path in self._unthemePaths:
            del self._unthemePaths[path]
        if path in self._usedPaths:
            del self._usedPaths[path]
        self._lock.release()
            		
    def classifySourceByTheme(self):
        wpsi = w.WallpaperServiceInfo()
        wpsi.getThemeInfo().classifyPathsByTheme( \
                self._unclassifiedPaths, \
                self._themePaths, \
                self._unthemePaths)
		
    def rewindSource(self):
        self._themeReady = False
        BPUtil.BPLog("rewindSource is called.")
        self._unclassifiedPaths = self._usedPaths
        self._usedPaths = OrderedDict()
        self.classifySourceByTheme()
        self._themeReady = True
		
    def prepareSourceForTheme(self):
        self._themeReady = False
        requestedThemeInfo = w.WallpaperServiceInfo().getThemeInfo()
        BPUtil.BPLog("prepareSourceForTheme is called with " + requestedThemeInfo._theme.stringValue())
        timer = Timer(1.1, self.prepareSourceForThemeInternal, [requestedThemeInfo])
        timer.start()
        
    def prepareSourceForThemeInternal(self, requestedThemeInfo):
        if (requestedThemeInfo.equals(w.WallpaperServiceInfo().getThemeInfo())):
            self._lock.acquire()
            self._unclassifiedPaths = self._themePaths
            backupPaths = self._unthemePaths
            self._themePaths = OrderedDict()
            self._unthemePaths = OrderedDict()
            self.classifySourceByTheme()
            self._unclassifiedPaths = backupPaths
            self.classifySourceByTheme()
            self._lock.release()
            self._themeReady = True
        BPUtil.BPLog("prepareSourceForTheme is done with " + requestedThemeInfo._theme.stringValue())
 
    def isThemeReady(self):
        return self._themeReady
		
    def getImageStat(self, path):
        self._lock.acquire()
        index = len(self._usedPaths) + len(self._unthemePaths)
        i = list(self._lastUsedPaths.keys()).index(path)
        if (i >= 0): # current image is in the mid offset
            index = index - len(self._lastUsedPaths) + i
            stat = str(index + 1) + " of " + str(self.getTotalImageCount())
        self._lock.release()
        return stat

    def getTotalImageCount(self):
        total = len(self._usedPaths) + len(self._themePaths) + len(self._unthemePaths)
        return total

    def readListCache(self):
        self._usingListCache = False
        self._unclassifiedPaths.clear()
        if (BPUtil.fileExists(_wallpaperListCacheFile)):
            self.waitForListFileWriting()
            with open(_wallpaperListCacheFile, mode = 'r', encoding = "utf-8") as f:
                path = f.readline().strip()
                if (path == self._sourceRootPath):
                    for line in f:
                        path_exif = line.strip().split('\t')
                        self._unclassifiedPaths[path_exif[0]] = ""
        if (BPUtil.fileExists(_photoListCacheFile)):
            self.waitForListFileWriting()
            with open(_photoListCacheFile, mode = 'r', encoding = "utf-8") as f:
                path = f.readline().strip()
                if (path == self._sourceRootPath):
                    for line in f:
                        path_exif = line.strip().split('\t')
                        exif =  path_exif[1] if len(path_exif) > 1 else ""
                        self._unclassifiedPaths[path_exif[0]] = exif
        self._usingListCache = len(self._unclassifiedPaths) > 0
        return self._usingListCache

    # None to delete
    def writeListCache(self, paths, listCacheFile):
        if (paths != None):
            self.waitForListFileWriting()
            self._writeFileThread = Thread(target = self.writeListCacheInternal, args = (paths, listCacheFile))
            self._writeFileThread.start()

    def writeListCacheInternal(self, paths, listCacheFile):
        BPUtil.BPLog("writeListCache", listCacheFile, "started")
        with open(listCacheFile + ".tmp", mode = 'w', encoding = "utf-8") as f:
            f.write(self._sourceRootPath + "\n")
            for key, value in paths.items():
                if (self._writeInterrupt):
                    BPUtil.BPLog("writeListCache is interrupted for", listCacheFile)
                    break
                wpath = WPath.WPath(key, value)
                f.write(wpath.path + "\t" + wpath.exif + "\n")
        if not self._writeInterrupt:
            if (BPUtil.fileExists(listCacheFile)):
                os.remove(listCacheFile)
            os.rename(listCacheFile + ".tmp", listCacheFile)
        else:
            os.remove(listCacheFile + ".tmp")
        BPUtil.BPLog("writeListCache", listCacheFile, "ended" )

    def waitForListFileWriting(self):
        if (self._writeFileThread != None) :
            BPUtil.BPLog("waitForListFileWriting started")
            self._writeFileThread.join()
            BPUtil.BPLog("waitForListFileWriting ended")

    def replaceCacheWithRealPaths(self, realPaths, realPaths2 = None):
        self._themeReady = False
        self._unclassifiedPaths = realPaths
        self._themePaths.clear()
        self._unthemePaths.clear()
        self.classifySourceByTheme()
        if realPaths2 != None:
            self._unclassifiedPaths = realPaths2
            self.classifySourceByTheme()
        for key, _ in self._usedPaths.items():
            # remove from _themePaths which was already used from cache
            # python must check key exists before del
            if key in self._themePaths:
                del self._themePaths[key]
        self._usingListCache = False
        self._themeReady = True
        WallpaperWidgetProvider.updateLabelWidget("Image list is updated.")

_wallpaperListCacheFile = ".wallpaper_list.txt"
_photoListCacheFile = ".photo_list.txt"

def getDefaultRootPath():
    return os.path.expanduser('~/Pictures') 
